// @wraith13/vscel/library/package.js
"use strict";
// ここの型情報は vscode.extensions.all.map(i => i.packageJSON) の為のモノで、これはオリジナルの package.json と微妙に異なっており
// 実際のデータとオリジナルの package.json のスキーマ情報( `https://json.schemastore.org/package` と `vscode://schemas/vscode-extensions` )
// を参考に書き起こした型情報となります。
Object.defineProperty(exports, "__esModule", { value: true });
exports.Category = exports.ThemeEnum = void 0;
var ThemeEnum;
(function (ThemeEnum) {
    ThemeEnum[ThemeEnum["dark"] = 0] = "dark";
    ThemeEnum[ThemeEnum["light"] = 1] = "light";
})(ThemeEnum = exports.ThemeEnum || (exports.ThemeEnum = {}));
var Category;
(function (Category) {
    Category[Category["Azure"] = 0] = "Azure";
    Category[Category["Data Science"] = 1] = "Data Science";
    Category[Category["Debuggers"] = 2] = "Debuggers";
    Category[Category["Extension Packs"] = 3] = "Extension Packs";
    Category[Category["Formatters"] = 4] = "Formatters";
    Category[Category["Keymaps"] = 5] = "Keymaps";
    Category[Category["Language Packs"] = 6] = "Language Packs";
    Category[Category["Linters"] = 7] = "Linters";
    Category[Category["Machine Learning"] = 8] = "Machine Learning";
    Category[Category["Notebooks"] = 9] = "Notebooks";
    Category[Category["Programming Languages"] = 10] = "Programming Languages";
    Category[Category["SCM Providers"] = 11] = "SCM Providers";
    Category[Category["Snippets"] = 12] = "Snippets";
    Category[Category["Themes"] = 13] = "Themes";
    Category[Category["Testing"] = 14] = "Testing";
    Category[Category["Visualization"] = 15] = "Visualization";
    Category[Category["Other"] = 16] = "Other";
})(Category = exports.Category || (exports.Category = {}));
